contrib/pg_tde/src/keyring/keyring_api.c

```diff
@@ -100,16 +100,55 @@ RegisterKeyProviderType(const TDEKeyringRoutine *routine, ProviderType type)
 KeyInfo *
 KeyringGetKey(GenericKeyring *keyring, const char *key_name, KeyringReturnCodes *returnCode)
 {
+	KeyInfo    *key = NULL;
 	RegisteredKeyProviderType *kp = find_key_provider_type(keyring->type);
 
 	if (kp == NULL)
 	{
 		ereport(WARNING,
-				errmsg("Key provider of type %d not registered", keyring->type));
+				errmsg("key provider of type %d not registered", keyring->type));
 		*returnCode = KEYRING_CODE_INVALID_PROVIDER;
 		return NULL;
 	}
-	return kp->routine->keyring_get_key(keyring, key_name, returnCode);
+	key = kp->routine->keyring_get_key(keyring, key_name, returnCode);
+
+	if (*returnCode != KEYRING_CODE_SUCCESS || key == NULL)
+		return NULL;
+
+	if (!ValidateKey(key))
+	{
+		*returnCode = KEYRING_CODE_INVALID_KEY;
+		pfree(key);
+		return NULL;
+	}
+
+	return key;
+}
+
+bool
+ValidateKey(KeyInfo *key)
+{
+	Assert(key != NULL);
+
+	if (key->name[0] == '\0')
+	{
+		ereport(WARNING, errmsg("invalid key: name is empty"));
+		return false;
+	}
+
+	if (key->data.len == 0)
+	{
+		ereport(WARNING, errmsg("invalid key: data length is zero"));
+		return false;
+	}
+
+	if (key->data.len > MAX_KEY_DATA_SIZE)
```

Comment:
We should validate exact key length, smaller also won't work properly.

Reply:
Also, if this happens, we already had a buffer overflow situation earlier, as we wrote more data into key->data.data than its capacity.

This validation should be earlier, inside the actual key provider code, before we write the data array.